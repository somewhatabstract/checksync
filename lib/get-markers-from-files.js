// @flow
/**
 * Cache in which we store our knowledge of existing markers.
 */

import uniq from "lodash/uniq";
import parseFile from "./parse-file.js";

import type {ILog, MarkerCache} from "./types.js";

/**
 * Generate a marker cache from the given files.
 *
 * @export
 * @param {Array<string>} files The files to be loaded.
 * @returns {Promise<MarkerCache>} A marker cache.
 */
export default async function getMarkersFromFiles(
    files: Array<string>,
    comments: Array<string>,
    log: ILog,
): Promise<MarkerCache> {
    const cacheData: MarkerCache = ({}: any);
    const referencedFiles: Array<string> = [];
    const logFileRef = fileRef => referencedFiles.push(fileRef);

    for (const file of files) {
        const fileMarkers = await parseFile(
            file,
            true,
            comments,
            log,
            logFileRef,
        );
        if (fileMarkers != null) {
            cacheData[file] = fileMarkers;
        }
        // Separate each file with a new line.
        log.log("");
    }

    /**
     * In this second pass, we load the files that are not in the original
     * set, and capture their markers. We don't care about going any level
     * deeper (though perhaps repeating till all referenced files are
     * loaded would be a mode folks might want).
     */
    for (const fileRef of uniq(referencedFiles)) {
        if (cacheData[fileRef]) {
            continue;
        }

        const fileMarkers = await parseFile(fileRef, false, comments, log);
        if (fileMarkers != null) {
            cacheData[fileRef] = fileMarkers;
        }
        // Separate each file with a new line.
        log.log("");
    }

    return cacheData;
}
