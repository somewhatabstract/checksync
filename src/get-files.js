// @flow
import glob from "fast-glob";
import path from "path";

import type {ILog} from "./types.js";

/**
 * Following gitignore format https://git-scm.com/docs/gitignore#_pattern_format
 *
 * /foo  Ignore root (not sub) file and dir and its paths underneath.     /foo, /foo/**
 * /foo/ Ignore root (not sub) foo dir and its paths underneath.          /foo/**
 * foo   Ignore (root/sub) foo files and dirs and their paths underneath. foo, ** /foo/**
 * foo/  Ignore (root/sub) foo dirs and their paths underneath.	          ** /foo/**
 */
function* turnIgnoresToGlobs(globs: Array<string>): Iterator<string> {
    const normalizeSeparators = (g: string): string =>
        g.split(path.sep).join("/");
    for (const glob of globs) {
        if (glob.startsWith("/")) {
            yield normalizeSeparators(path.join(glob, "**"));
            if (!glob.endsWith("/")) {
                yield glob;
            }
        } else {
            yield normalizeSeparators(path.join("**", glob, "**"));
            if (!glob.endsWith("/")) {
                yield glob;
            }
        }
    }
}

/**
 * Expand the given globs into files.
 *
 * @param {Array<string>} globs The globs to expand.
 */
export default async function getFiles(
    includeGlobs: Array<string>,
    excludeGlobs: Array<string>,
    log: ILog,
): Promise<Array<string>> {
    const includePatterns = Array.from(turnIgnoresToGlobs(includeGlobs));
    log.verbose(
        () => `Include globs: ${JSON.stringify(includePatterns, null, "    ")}`,
    );
    const excludePatterns = Array.from(turnIgnoresToGlobs(excludeGlobs));
    log.verbose(
        () => `Exclude globs: ${JSON.stringify(excludePatterns, null, "    ")}`,
    );

    // Now let's match the patterns and see what files we get.
    const paths = await glob(includePatterns, {
        onlyFiles: true,
        absolute: true,
        ignore: excludePatterns,
    });
    const sortedPaths = paths
        .map((p) => p.replace(new RegExp("/", "g"), path.sep))
        .sort();
    log.verbose(
        () => `Discovered paths: ${JSON.stringify(sortedPaths, null, "    ")}`,
    );
    return sortedPaths;
}
