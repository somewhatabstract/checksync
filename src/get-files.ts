import glob from "fast-glob";
import path from "path";
import fs from "fs";
import ignoreFileGlobsToAllowPredicate from "./ignore-file-globs-to-allow-predicate";

import {ILog, Options} from "./types";

/**
 * Expand the given globs and ignore files into a list of files.
 *
 * @param {Array<string>} includeGlobs The include globs to expand.
 * @param {Array<string>} excludeGlobs The exclude globs to expand.
 * @param {Array<string>} ignoreFiles The ignore files to expand.
 * @param {ILog} log A log to record things
 */
export default async function getFiles(
    {includeGlobs, excludeGlobs, ignoreFiles, includeDotPaths}: Options,
    log: ILog,
): Promise<Array<string>> {
    // We turn all the ignore files we're using into a single predicate that
    // returns true if a file is allowed, or false if it is ignored.
    const allowPredicate = await ignoreFileGlobsToAllowPredicate(
        ignoreFiles,
        log,
    );

    // If any of our input globs don't contain a * and they are a directory
    // then let's make that be a glob of everything under that dir.
    // If we don't do this then checksync feels counterintuitive as
    // just a directory like __examples__ won't work, and __examples__/**
    // may get expanded by the shell, so folks would have to do
    // __examples__/**/* which is just annoying.
    // With this change, any input directories are expanded to mean all
    // files below that directory and that directory's children.
    includeGlobs = includeGlobs.map((pattern: string) =>
        !pattern.includes("*") &&
        // TODO: Use throwIfNoEntry in lstatSync/lstat to replace need to do
        // the existence check, once we are on versions of Node that support it.
        fs.existsSync(pattern) &&
        fs.lstatSync(pattern).isDirectory()
            ? `${pattern}/**`
            : pattern,
    );

    log.verbose(
        () => `Include globs: ${JSON.stringify(includeGlobs, null, 4)}`,
    );
    log.verbose(
        () => `Exclude globs: ${JSON.stringify(excludeGlobs, null, 4)}`,
    );

    // Now let's match the patterns and see what files we get.
    const paths = await glob([...includeGlobs], {
        onlyFiles: true,
        absolute: true,
        dot: includeDotPaths,
        ignore: excludeGlobs as Array<string>, // remove readonly-ness
    });
    const sortedPaths = paths
        // Sort the paths first to ensure we get a deterministic output.
        .sort()
        // Filter out any paths that are not allowed by the ignore files.
        .filter((p) => allowPredicate(p))
        // Replace the `/` with the OS-specific path separator.
        .map((p) => p.replace(new RegExp("/", "g"), path.sep));
    log.verbose(
        () => `Discovered paths: ${JSON.stringify(sortedPaths, null, 4)}`,
    );
    return sortedPaths;
}
