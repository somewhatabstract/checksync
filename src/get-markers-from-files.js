// @flow
/**
 * Cache in which we store our knowledge of existing markers.
 */
import fs from "fs";
import path from "path";
import uniq from "lodash/uniq";
import parseFile from "./parse-file.js";

import type {ILog, MarkerCache} from "./types.js";

/**
 * Generate a marker cache from the given files.
 *
 * @export
 * @param {Array<string>} files The files to be loaded.
 * @returns {Promise<MarkerCache>} A marker cache.
 */
export default async function getMarkersFromFiles(
    files: Array<string>,
    comments: Array<string>,
    log: ILog,
): Promise<MarkerCache> {
    const cacheData: MarkerCache = {};
    const referencedFiles: Array<string> = [];
    const logFileRef = (file, fileRef) => {
        const normalizedFileRef = path.resolve(path.dirname(file), fileRef);
        const exists =
            fs.existsSync(normalizedFileRef) &&
            fs.lstatSync(normalizedFileRef).isFile();
        if (exists) {
            referencedFiles.push(normalizedFileRef);
        }
        return {file: normalizedFileRef, exists};
    };

    // TODO(somewhatabstract): Use jest-worker and farm parsing out to
    // multiple threads.

    for (const file of files) {
        const fileMarkers = await parseFile(
            file,
            true,
            comments,
            log,
            fileRef => logFileRef(file, fileRef),
        );
        cacheData[file] = fileMarkers;
    }

    /**
     * In this second pass, we load the files that are not in the original
     * set, and capture their markers. We don't care about going any level
     * deeper (though perhaps repeating till all referenced files are
     * loaded would be a mode folks might want).
     */
    for (const fileRef of uniq(referencedFiles)) {
        if (cacheData[fileRef] !== undefined) {
            continue;
        }

        const fileMarkers = await parseFile(fileRef, false, comments, log);
        cacheData[fileRef] = fileMarkers;
    }

    return cacheData;
}
