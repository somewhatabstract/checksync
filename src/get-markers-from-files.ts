/**
 * Cache in which we store our knowledge of existing markers.
 */
import fs from "fs";
import parseFile from "./parse-file";

import {FileInfo, MarkerCache, Options} from "./types";
import * as Errors from "./errors";

/**
 * Generate a marker cache from the given files.
 *
 * @export
 * @param {Options} options
 * @param {Array<string>} files The files to be loaded.
 * @returns {Promise<MarkerCache>} A marker cache.
 */
export default async function getMarkersFromFiles(
    options: Options,
    files: ReadonlyArray<string>,
): Promise<MarkerCache> {
    const cacheData: MarkerCache = {};
    const referencedFiles: Array<string> = [];

    const setCacheData = (file: string, info: FileInfo) => {
        cacheData[file] = info;
        info.aliases.push(file);
    };

    const cacheFiles = async (
        files: ReadonlyArray<string>,
        readOnly: boolean,
    ) => {
        for (const file of files) {
            let realFilePath;
            try {
                // This could be a symlink.
                // We want to treat it as the canonical real file.
                realFilePath = fs.realpathSync(file);

                // We don't need to parse things twice.
                if (cacheData[realFilePath] === undefined) {
                    // OK, it hasn't been parsed yet. We need to parse it so we
                    // can store a copy for our symlink.
                    const parseResult = await parseFile(
                        options,
                        realFilePath,
                        readOnly,
                    );
                    setCacheData(realFilePath, {
                        readOnly,
                        markers: parseResult.markers || {},
                        aliases: [],
                        errors: parseResult.errors,
                        lineCount: parseResult.lineCount,
                    });
                    referencedFiles.push(...parseResult.referencedFiles);
                }
            } catch (e: any) {
                // If we got an issue, then we need to store this as could
                // not parse (as well as record the unfixable alias)
                setCacheData(file, {
                    readOnly,
                    markers: {},
                    aliases: [],
                    errors: [Errors.couldNotParse(file, e.message)],
                });
            }

            // If the original file was a symlink, let's make sure we
            // store the markers (or the error) under its filepath too.
            if (realFilePath && realFilePath !== file) {
                // Close as unfixable, since this file already exists in
                // its potentially fixable version, and we don't need to fix
                // it twice.
                setCacheData(file, {
                    ...(cacheData[realFilePath] ?? {}),
                    readOnly: true,
                });
            }
        }
    };

    // Process the main file set. These are considered "fixable" as they
    // are the files that our user requested be processed directly so we
    // pass readOnly as false.
    await cacheFiles(files, false);

    /**
     * In this second pass, we load the files that were referenced by the
     * files we just processed and capture their markers.
     */
    await cacheFiles(referencedFiles, true);

    return cacheData;
}
