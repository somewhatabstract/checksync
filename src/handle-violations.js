// @flow
import fixViolation from "./fix-violation.js";
import reportViolation from "./report-violation.js";

import type {MarkerCache, ILog} from "./types";

export default function handleViolations(
    cache: MarkerCache,
    autoFix: boolean,
    log: ILog,
): Array<string> {
    const violationHandler = autoFix ? fixViolation : reportViolation;

    const violationFiles = {};
    for (const file of Object.keys(cache)) {
        const markers = cache[file];
        if (markers == null) {
            // This means a target reference that couldn't be found and we can
            // totally ignore it at this level.
            continue;
        }

        for (const markerID of Object.keys(markers)) {
            const marker = markers[markerID];
            if (!marker.fixable) {
                // If this marker isn't one that can be fixed, we don't care
                // if it is a violation or not; it was only loaded for reference
                // as a target of a fixable marker.
                continue;
            }

            for (const line of Object.keys(marker.targets)) {
                const lineNumber = parseInt(line);
                const targetRef = marker.targets[lineNumber];

                const target = cache[targetRef.file];
                if (target == null) {
                    // Target doesn't exist if we get null.
                    // We already indicated this error elsewhere, so just skip
                    // along.
                    continue;
                }
                const targetMarker = target[markerID];

                // Now compare the actual target marker checksum with the one
                // in our reference, and if they don't match, do something
                // about it.
                if (targetMarker.checksum !== targetRef.checksum) {
                    violationFiles[file] = true;
                    violationHandler(
                        file,
                        lineNumber,
                        markerID,
                        targetRef.checksum,
                        targetRef.file,
                        targetMarker.checksum,
                        marker.fixable,
                        log,
                    );
                }
            }
        }
    }
    return Object.keys(violationFiles);
}
