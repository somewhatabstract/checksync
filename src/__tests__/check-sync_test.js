// @flow
import path from "path";
import chalk from "chalk";
import * as GetFiles from "../get-files.js";
import * as GetMarkersFromFiles from "../get-markers-from-files.js";
import * as HandleViolations from "../handle-violations.js";
import Logger from "../logger.js";
import StringLogger from "../string-logger.js";

import checkSync from "../check-sync.js";
import ErrorCodes from "../error-codes.js";

jest.mock("../get-files.js");
jest.mock("../get-markers-from-files.js");
jest.mock("../handle-violations.js");

describe("#checkSync", () => {
    beforeAll(() => {
        chalk.enabled = false;
    });

    afterAll(() => {
        chalk.enabled = true;
    });

    it("should expand the globs to files", async () => {
        // Arrange
        const NullLogger = new Logger();
        const getFilesSpy = jest.spyOn(GetFiles, "default").mockReturnValue([]);

        // Act
        await checkSync(["glob1", "glob2"], true, ["//"], NullLogger);

        // Assert
        expect(getFilesSpy).toHaveBeenCalledWith(["glob1", "glob2"]);
    });

    it("should log error when there are no matching files", async () => {
        // Arrange
        const NullLogger = new Logger();
        jest.spyOn(GetFiles, "default").mockReturnValue([]);
        const errorSpy = jest.spyOn(NullLogger, "error");

        // Act
        await checkSync(["glob1", "glob2"], false, ["//"], NullLogger);

        // Assert
        expect(errorSpy).toHaveBeenCalledWith("No matching files");
    });

    it("should return NO_FILES when there are no matching files", async () => {
        // Arrange
        const NullLogger = new Logger();
        jest.spyOn(GetFiles, "default").mockReturnValue([]);

        // Act
        const result = await checkSync(
            ["glob1", "glob2"],
            false,
            ["//"],
            NullLogger,
        );

        // Assert
        expect(result).toBe(ErrorCodes.NO_FILES);
    });

    it("should build a marker cache from the files", async () => {
        // Arrange
        const NullLogger = new Logger();
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(HandleViolations, "default").mockReturnValue([]);
        const getMarkersFromFilesSpy = jest
            .spyOn(GetMarkersFromFiles, "default")
            .mockReturnValue({});

        // Act
        await checkSync([], true, ["//"], NullLogger);

        // Assert
        expect(getMarkersFromFilesSpy).toHaveBeenCalledWith(
            ["filea", "fileb"],
            ["//"],
            NullLogger,
        );
    });

    it("should log error if there were errors during cache build", async () => {
        // Arrange
        const NullLogger = new Logger();
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        const errorSpy = jest.spyOn(NullLogger, "error");
        jest.spyOn(GetMarkersFromFiles, "default").mockImplementation(() => {
            NullLogger.error("Oh no!");
            return {};
        });

        // Act
        await checkSync([], true, ["//"], NullLogger);

        // Assert
        expect(errorSpy).toHaveBeenCalledWith("Aborting due to errors");
    });

    it("should return PARSE_ERRORS when there are no matching files", async () => {
        // Arrange
        const NullLogger = new Logger();
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(GetMarkersFromFiles, "default").mockImplementation(() => {
            NullLogger.error("Oh no!");
            return {};
        });

        // Act
        const result = await checkSync([], true, ["//"], NullLogger);

        // Assert
        expect(result).toBe(ErrorCodes.PARSE_ERRORS);
    });

    it("should invoke handleViolations with cache, autoFix, and log", async () => {
        // Arrange
        const NullLogger = new Logger();
        const fakeCache = {};
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(GetMarkersFromFiles, "default").mockReturnValue(fakeCache);
        const handleViolationsSpy = jest
            .spyOn(HandleViolations, "default")
            .mockReturnValue([]);

        // Act
        await checkSync([], false, ["//"], NullLogger);

        // Assert
        expect(handleViolationsSpy).toHaveBeenCalledWith(
            fakeCache,
            false,
            NullLogger,
        );
    });

    it("should return SUCCESS if no violations", async () => {
        // Arrange
        const NullLogger = new Logger();
        const fakeCache = {};
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(GetMarkersFromFiles, "default").mockReturnValue(fakeCache);
        jest.spyOn(HandleViolations, "default").mockReturnValue([]);

        // Act
        const result = await checkSync([], false, ["//"], NullLogger);

        // Assert
        expect(result).toBe(ErrorCodes.SUCCESS);
    });

    it("should return SUCCESS if violations with autoFix", async () => {
        // Arrange
        const NullLogger = new Logger();
        const fakeCache = {};
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(GetMarkersFromFiles, "default").mockReturnValue(fakeCache);
        jest.spyOn(HandleViolations, "default").mockReturnValue(["violation"]);

        // Act
        const result = await checkSync([], true, ["//"], NullLogger);

        // Assert
        expect(result).toBe(ErrorCodes.SUCCESS);
    });

    it("should output summary if violations with autoFix", async () => {
        // Arrange
        const NullLogger = new Logger();
        const fakeCache = {};
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(GetMarkersFromFiles, "default").mockReturnValue(fakeCache);
        jest.spyOn(HandleViolations, "default").mockReturnValue(["violation"]);
        const infoSpy = jest.spyOn(NullLogger, "info");

        // Act
        await checkSync([], true, ["//"], NullLogger);

        // Assert
        expect(infoSpy).toHaveBeenCalledWith("Fixed 1 file(s)");
    });

    it("should return DESYNCHRONIZED_BLOCKS if violations without autoFix", async () => {
        // Arrange
        const NullLogger = new Logger();
        const fakeCache = {};
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(GetMarkersFromFiles, "default").mockReturnValue(fakeCache);
        jest.spyOn(HandleViolations, "default").mockReturnValue(["violation"]);

        // Act
        const result = await checkSync([], false, ["//"], NullLogger);

        // Assert
        expect(result).toBe(ErrorCodes.DESYNCHRONIZED_BLOCKS);
    });

    it("should example of how to fix violations if not autoFix", async () => {
        // Arrange
        const NullLogger = new Logger();
        const fakeCache = {};
        jest.spyOn(GetFiles, "default").mockReturnValue(["filea", "fileb"]);
        jest.spyOn(GetMarkersFromFiles, "default").mockReturnValue(fakeCache);
        jest.spyOn(HandleViolations, "default").mockReturnValue([
            "violation1",
            "violation2",
        ]);
        const errorSpy = jest.spyOn(NullLogger, "error");

        // Act
        await checkSync([], false, ["//"], NullLogger);

        // Assert
        expect(errorSpy).toHaveBeenCalledWith(
            "checksync --fix violation1 violation2",
            true,
        );
    });

    it("should report errors properly for __examples__", async () => {
        // Arrange
        jest.spyOn(GetFiles, "default").mockImplementation((...args) =>
            jest.requireActual("../get-files.js").default(...args),
        );
        jest.spyOn(HandleViolations, "default").mockImplementation((...args) =>
            jest.requireActual("../handle-violations.js").default(...args),
        );
        jest.spyOn(GetMarkersFromFiles, "default").mockImplementation(
            (...args) =>
                jest
                    .requireActual("../get-markers-from-files.js")
                    .default(...args),
        );
        const stringLogger = new StringLogger();

        // Act
        await checkSync(
            [path.join(__dirname, "../../__examples__")],
            false,
            ["//", "#"],
            stringLogger,
        );
        const result = stringLogger.getLog();

        // Assert
        expect(result).toMatchSnapshot("__examples__");
    });
});
