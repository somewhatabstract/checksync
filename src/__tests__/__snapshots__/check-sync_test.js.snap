// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#checkSync should report errors properly for __examples__: __examples__ 1`] = `
"<group> INFO  /Users/jeffyates/git/checksync/__examples__/content_after_start/c.js
   ERROR  Sync-tag \\"content_after_start\\" target found after content started
<end_group>
<group> INFO  /Users/jeffyates/git/checksync/__examples__/directory_target/example.js
   ERROR  Sync-tag \\"directory_target\\" points to \\"./a_directory\\", which does not exist or is a directory
<end_group>
<group> INFO  /Users/jeffyates/git/checksync/__examples__/end_with_no_start/example.js
   ERROR  Sync-tag \\"end_with_no_start\\" end found, but sync-tag never started
<end_group>
<group> INFO  /Users/jeffyates/git/checksync/__examples__/missing_target/example.js
   ERROR  Sync-tag \\"missing_target\\" points to \\"./missing_target.py\\", which does not exist or is a directory
<end_group>
<group> INFO  /Users/jeffyates/git/checksync/__examples__/no_self_reference/example.js
   ERROR  Sync-tag \\"example_three\\" cannot target itself
<end_group>
<group> INFO  /Users/jeffyates/git/checksync/__examples__/unterminated_marker/example_two-b.py
  <group> ERROR  Unterminated markers
    example_two
  <end_group>
<end_group>
 ERROR  Aborting due to errors"
`;
